Mathlib/LinearAlgebra/Matrix/Determinant/TotallyUnimodular.lean

```diff
@@ -187,4 +188,134 @@ lemma fromColumns_col0_isTotallyUnimodular_iff (A : Matrix m n R) :
   rw [← transpose_isTotallyUnimodular_iff, transpose_fromColumns, transpose_col,
     fromRows_row0_isTotallyUnimodular_iff, transpose_isTotallyUnimodular_iff]
 
+-- TODO maybe rename and certainly move
+lemma in_set_range_singType_cast_mul_in_set_range_singType_cast {a b : R}
+    (ha : a ∈ Set.range SignType.cast) (hb : b ∈ Set.range SignType.cast) :
+    a * b ∈ Set.range SignType.cast := by
+  sorry
+
+variable {R : Type*} [LinearOrderedCommRing R] -- TODO reorganize
+
+-- TODO maybe rename and certainly move
+lemma in_set_range_singType_cast_iff_abs (a : R) :
+    a ∈ Set.range SignType.cast ↔ |a| ∈ Set.range SignType.cast := by
+  sorry
+
+-- TODO maybe rename and certainly move; do not require `R` to be algebraic structure
+lemma fromBlocks_submatrix (A₁ : Matrix m n R) (A₂ : Matrix m' n' R)
+    {α : Type*} (f : α → m ⊕ m') (g : α → n ⊕ n') :
+    (fromBlocks A₁ 0 0 A₂).submatrix f g =
+    (Matrix.fromBlocks
+      (A₁.submatrix
+        ((·.val.snd) : { x₁ : α × m // f x₁.fst = Sum.inl x₁.snd } → m)
+        ((·.val.snd) : { y₁ : α × n // g y₁.fst = Sum.inl y₁.snd } → n)
+      ) 0 0
+      (A₂.submatrix
+        ((·.val.snd) : { x₂ : α × m' // f x₂.fst = Sum.inr x₂.snd } → m')
+        ((·.val.snd) : { y₂ : α × n' // g y₂.fst = Sum.inr y₂.snd } → n')
+      )
+     ).submatrix f.decomposeSum g.decomposeSum := by
+  rw [f.eq_comp_decomposeSum, g.eq_comp_decomposeSum, ←Matrix.submatrix_submatrix]
+  aesop
+
+attribute [-simp] Fintype.card_ofIsEmpty Fintype.card_ofSubsingleton -- major performance issue
+
+lemma fromBlocks_isTotallyUnimodular_zero_zero_isTotallyUnimodular
+    [Fintype m] [DecidableEq m] [Fintype m'] [DecidableEq m']
+    [Fintype n] [DecidableEq n] [Fintype n'] [DecidableEq n']
+    {A₁ : Matrix m n R} {A₂ : Matrix m' n' R}
+    (hA₁ : A₁.IsTotallyUnimodular) (hA₂ : A₂.IsTotallyUnimodular) :
+    (fromBlocks A₁ 0 0 A₂).IsTotallyUnimodular := by
+  intro k f g hf hg
+  rw [isTotallyUnimodular_iff] at hA₁ hA₂
+  rw [fromBlocks_submatrix]
+  -- look at sizes of submatrices in blocks
+  if hxy : Fintype.card { x₁ : Fin k × m // f x₁.fst = Sum.inl x₁.snd }
```

Comment:
Would the remainder of the proof simplify if you ask for `Nonempty (X \equiv Y)`, where I use `X` and `Y` as abbrevs for those subtypes?

Reply:
What do you mean by "ask" here? Is it about being as the `if` condition instead of asking for size equality?

Reply:
If I understand correctly what you wrote, it would not simplify the proof, but the code might be a bit shorter. In any case, this is rather a microöptimisation that I wouldn't do now.

Reply:
After the new change, it no longer makes a difference.